luci-base: http.uc: add CSP headers to http response

josteink opened the pull request:
- [X] This PR is not from my *main* or *master* branch :poop:, but a *separate* branch :white_check_mark:
- [X] Each commit has a valid :black_nib: `Signed-off-by: <my@email.address>` row (via `git commit --signoff`)
- [X] Each commit and PR title has a valid :memo: `<package name>: title` first line subject for packages
- [X] Incremented :up: any `PKG_VERSION` in the Makefile
- [X] Tested on: (Atheros AR9132 rev 2, 25.12.0-RC1, Chromium) :white_check_mark:
- [X] \( Preferred ) Mention: @josteink
- [X] \( Preferred ) Screenshot or mp4 of changes: 
<img width="856" height="228" alt="image" src="https://github.com/user-attachments/assets/802d0f81-662b-4972-97b2-1fab7868dcd9" />
<img width="1920" height="1080" alt="image" src="https://github.com/user-attachments/assets/5755d869-ce35-4e4f-9399-f93f6bcbf5cb" />

- [X] \( Optional ) Closes: e.g. openwrt/luci#8160
- [X] Description: (describe the changes proposed in this PR)

This PR adds support for Content Security Policy to Luci Web.

Content Security Policy is a very effective security hardening for browser-based applications:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Guides/CSP

CSP for OpenWRT/Luci is implemented in 4 different modes:

- None (Disabled - least secure) - default
- Permissive (More secure. Should permit extensive customization)
- Strict (Most secure. Only allow what is required by OpenWRT/Luci by default) - block everything else
- Custom - allow user to define own CSP policy

This option is stored in `/etc/config/uhttpd` and also manageable via the Luci Web UI under "System" -> "Administration" -> "uHTTPd" -> "HTTP(s) Acces" -> "General settings".

josteink committed:
luci-base: http.uc: add CSP HTTP response headers

This commit adds support for Content Security Policy to Luci Web.

Content Security Policy is a very effective security hardenining
for browser-based applications.

https://developer.mozilla.org/en-US/docs/Web/HTTP/Guides/CSP

The CSP can be enabled/disabled and customized via uci.

The default "strict" provided CSP effectively limits the Luci web-app
to only allow:

- scripts from its own origin/IP, and nowhere else.
- css from its own origin/IP, and nowhere else.
- images from its own origin/IP, and nowhere else.
- fetch/etc from own origin/IP. and sysupgrade.
  nowhere else.
- (etc)

Softening exceptions:

- Allow javascript to use inline scripts, eval() and trusted types.
- Allow CSS to use inline styles.
- Allow IMG-tags with data: and blob:-type URLs.
- Allow XHR/fetch for any https:// url - required to support attended
  sysupgrade.

trusted-types-eval is required by Safari on MacOS, supported in
Firefox Nightly, and ignored in Chrome/Chromium.

This CSP should prevent almost any kind of XSS or script-injection
while at the same time allow Luci to perform any task it needs to do.

The default "permissive" allows more https:// sources.

"Custom" is defined by the user entirely.

Signed-off-by: Jostein Kjønigsen <jostein@kjonigsen.net>

josteink commented:
With the testing I've been able to to, it seems like attended sysupgrade now works as well!

efahl commented:
Hmm, maybe another one.  I have luci-app-statistics installed on one of my APs, and am getting this when I try to view the generated images:

```
Content-Security-Policy: The page’s settings blocked the loading of a resource (img-src) at blob:https://wap01/0bd0c7d6-fc6c-4917-a773-8847d822c46e because it violates the following directive: “img-src 'self' data:”
```
Hacked it up to contain `blob:` and it works:
```
507                 if (!this.headers?.['Content-Security-Policy'])
508                         this.header('Content-Security-Policy',
509                         "default-src 'self'; script-src 'self' 'unsafe-inline' 'unsafe-eval' 'trusted-types-eval';"
510                         + "img-src 'self' data: blob:;"
511                         + "style-src 'self' 'unsafe-inline';"
512                         + "connect-src 'self' https://*;");
```

efahl commented:
And firefox says
```
Content-Security-Policy: Couldn’t parse invalid host 'trusted-types-eval'
```
But how do browsers deal with these "bad" ones?  Just ignore them?  In which case, who cares if it "errors"?

That one specifically seems to have very narrow support, but if it is supported, seems like it should be there.

https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Content-Security-Policy/script-src#browser_compatibility

systemcrash commented:
For reasons you're now discovering, this is why we've not bothered with extended security measures like this, since most of the customisation people do or use depends on the freedom to do so. There are perhaps only a few apps which call to external resources like this, and the sources are entirely verifiable. Introducing this change will likely break some functionality somewhere... :/

josteink commented:
> For reasons you're now discovering, this is why we've not bothered with extended security measures like this, since most of the customisation people do or use depends on the freedom to do so.

These are IMO simple issues to work around or loosen up. 

If people are qualified to customize these things, they’re probably qualified to update or remove this header too? 

> There are perhaps only a few apps which call to external resources like this

People might get malicious browser extensions which wants to inject scripts into your router interface.

This would help protect against that to a certain extent. 

> and the sources are entirely verifiable. Introducing this change will likely break some functionality somewhere... :/

IMO script-src is the most important one to get in and the rest can be loosened where there’s any plausible reason for doing so.

Ramon00 commented:
It could be made configurable indeed. Add a option somewhere maybe in the system menu or something, different degrees of allowing things. Add a help text to explain.

mdevolde commented:
> could contain a config-pane

Or just a text area to write your CSP from start to finish, which is injected into `http.uc` (this way, anyone who wants to apply a CSP can completely customise their CSP according to their config).

Ramon00 commented:
> I think that's OK for a version 1 implementation, isn't it?

Depends on the default, if the default is empty then this is NOK (people would not know what to fill in).

I do like the idea of user defined. 

I would still go for drop down, 1) none 2) loose 3) tight 4) user defined. {previewing the new header) This is most user friendly and will serve everybody. I'm guessing 99% of the people will use 1), 2) or 3).

Do not choose for the easy solution just because you do not have the experience to implement, there are plenty of people willing to help out.

I do wonder what the default should be, 1) or 2) or 3)...

josteink committed:
luci-app-uhttpd: add configuration options for CSP

- add dropdown to select CSP-mode
- add textbox to configure custom CSP policy

Signed-off-by: Jostein Kjønigsen <jostein@kjonigsen.net>

systemcrash commented on applications/luci-app-uhttpd/htdocs/luci-static/resources/view/uhttpd/uhttpd.js:
```diff
@@ -96,6 +96,15 @@ return view.extend({
+		o.default = o.enabled;
+		o.rmempty = false;
+
+		o = s.taboption('general', form.Value, 'csp_policy', _('CSP Policy'), _('Custom Content-Security-Policy header value. Leave empty to use the default policy.'));
```
Use a textbox.

josteink commented on modules/luci-base/ucode/http.uc:
```diff
@@ -504,6 +506,24 @@ const Class = {
+		if (!this.headers?.['Content-Security-Policy']) {
+			// Read UCI settings - default to enabled with default policy
+			let uci_cursor = cursor();
+			let enable_csp = uci_cursor.get('uhttpd', 'main', 'enable_csp');
```
Fixed as originally suggested by @systemcrash .

josteink commented:
> It actually starts with a space before "luci-app-uhttpd", if you closely to the check results.

Got it. Fixed now!

> @Ramon00 wrote:
>
> I would still go for drop down, 1) none 2) loose 3) tight 4) user defined. {previewing the new header) This is most user friendly and will serve everybody. I'm guessing 99% of the people will use 1), 2) or 3).

I instead went with "None", "Permissive", "Strict" & "Custom", but now there's dropdowns there 😄 

I'll test more thoroughly tomorrow to ensure there's no stupid bugs, but any comments on the "form" of the UI solution would be appreciated.

mdevolde commented on applications/luci-app-uhttpd/htdocs/luci-static/resources/view/uhttpd/uhttpd.js:
```diff
@@ -96,6 +96,28 @@ return view.extend({
+		csp_mode_option.rmempty = false;
+
+		var csp_policy_option = s.taboption('general', form.Value, 'csp_policy', _('Custom CSP Policy String'), _('The Content-Security-Policy header-value used in custom-mode.') + "<br />" + _(' WARNING: Wrong values for this setting can render the web-UI inaccessible and require recovery by SSH.'));
+		csp_policy_option.default = 'default-src \'none\'; script-src \'self\' \'unsafe-inline\' \'unsafe-eval\' \'trusted-types-eval\'; img-src \'self\' data: blob:; style-src \'self\' \'unsafe-inline\'; connect-src \'self\' https://sysupgrade.openwrt.org;'
```
@systemcrash ping

josteink commented:
> The problems I see at the moment are as follows:
> 
> * settings in an app that might not be installed

I may be misunderstanding something, but the file I modified (uhttpd.js) for the settings is where these kind of HTTP controls used to be already.

> This potentially introduces new behaviour, yet does not surface a control element. 

This only introduces a new behaviour, if configured. What control elements are you looking for, beside the UI provided?

> The behaviour is in base, but the config element is in a separate app. luci-app-uhttpd seems like a good place for it. But luci-app-uhttpd configures uhttpd behaviour. This change is about controlling the front-end behaviour. It seems to me the better placement would be under system->admin->http, which is about the only decent place in base.

The setting you are referring to seems (to my untrained eye) to be uhttpd.js. Check the URL for the admin-page you are referring to:

<img width="1249" height="437" alt="image" src="https://github.com/user-attachments/assets/d6164e9b-4dca-4534-b031-d640b9b5cd8c" />

When grepping the source for "Redirect to HTTPS", this setting is only found in `uhttpd.js`.

```
% git grep "Redirect all HTTP to HTTPS" | grep .js
applications/luci-app-uhttpd/htdocs/luci-static/resources/view/uhttpd/uhttpd.js:		o = s.taboption('general', form.Flag, 'redirect_https', _('Redirect all HTTP to HTTPS'));
````

> This addition correctly 'fails open' - if there is no config present, it remains in a disabled state. You might want to flag that in the GUI - that disabled is the default state.

Good idea. I'll make it so.

> It also might be a bit inconsistent, but use " to wrap the string, and ' to quote the elements in it.

Fine by me :smile:

> What settings can render the UI inaccessible?

Only setting which can render the UI inaccessible is "custom" with a bad policy which excludes required JS scripts, in the area clearly marked as "experts only", and with a safe default provided.

**Edit:** I see my grep was incorrect and there is a different module I can (should?) insert these changes into.

I'll make it so too!

@systemcrash : I assume `modules/luci-mod-system/htdocs/luci-static/resources/view/system/uhttpd.js` is the right file then, to avoid dependency on non-default app? :) 

Or should I keep it in both?

josteink committed:
luci-mod-system: add configuration options for CSP

- add dropdown to select CSP-mode
- add textbox to configure custom CSP policy

Signed-off-by: Jostein Kjønigsen <jostein@kjonigsen.net>

josteink commented:
@systemcrash : I tried the following code to just to test. It works, kinda, but I always get the validate-handler invoked twice on changes, leading to the ui popup to appear twice as well.

```js
		csp_mode_option.validate = function(section_id, value) {
			// triggers UI warning when using custom-mode with UI timer based fallback
			if (value === "custom")
			{
			  ui.addTimeLimitedNotification(_('Danger zone'), E('p', _('Custom CSP header will be used as is. If deemed invalid by the browser, you will need to use SSH to reset the value in /etc/config/uhttpd.')), 60000, 'warning');
			}
			return true;
		};
```

<img width="1249" height="579" alt="image" src="https://github.com/user-attachments/assets/7bbc0fd7-832b-4ca6-841f-73e385df2c28" />

Is this a general luci problem, or is it anything else I'm doing wrong? 😄

systemcrash commented:
> Is there a convention somewhere on how to avoid double warnings then, or should I just leave it?

There is, IIRC. I just can't remember it now... let me have a think. 


Ah, sorry - take a look at: https://openwrt.github.io/luci/jsapi/LuCI.form.TextValue.html

There are some in use in the code-base. Take a peek at how others are used. 

> That is, you use the other policies to whitelist only the things you know you need.

I took a look [at MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Content-Security-Policy#fetch_directive_syntax) which says:

the single value `'none'`, indicating that the specific resource type should be completely blocked

So we can at least add a note about that in the GUI: e.g. `'none' completely blocks the specific resource type`, and if the user saws off the branch they're standing on, it's on them :)

josteink commented:
No new changes in latest pull. Have been busy during Christmas as everyone else 😄

Just did a rebase on upstream git master to ensure I'm not walking into any merge-conflicts.

systemcrash commented on modules/luci-base/ucode/http.uc:
```diff
@@ -504,6 +508,33 @@ const Class = {
 			this.header('X-Content-Type-Options', 'nosniff');
 
+		if (!this.headers?.['Content-Security-Policy']) {
+			// Read UCI settings
```
This will assign it every time - because each of `this.headers` gets squashed to lowercase in the addition function:
```js

	header: function(key, value) {
		this.headers ??= {};
		this.headers[lc(key)] = value;
	},
```

josteink replied:
Good call. I didn't know that. Fixed!

systemcrash commented:
I'm thinking a structure like: plugins -> class -> type, so here plugins -> http -> headers

And we progressively build the minimal changes where they need to go. Like for headers as discussed above. 

I'm working on a dynamic loading of plugin config pages. Install adds config page, plugin page displays it. Click - done. System (http) checks preference for class (on/off) and type (enabled/disabled) and loads. 

Something like that.

Config I think would go in luci -> plugin (section type).

josteink commented:
Sorry for the long wait. I've been busy across the line.

I've now gotten a spare device, provisioned with OpenWRT SNAPSHOT, so that I can test the plugin system.

Due to how everything is changed (and this PR is with a certain degree of bike shedding)... I'll leave this PR be, and start a new one:

https://github.com/openwrt/luci/pull/8645

systemcrash commented:
Closed via #8645
